### PR TITLE
Various gctx gl improvements

### PR DESCRIPTION
**Pull request by mbouron-gpsw**
The goal of this PR is to remove the use of NULL as a meaning to provide the render pass associated with the OpenGL default framebuffer. This is a first step towards moving the begin and end of the default renderpass outside of begin_draw()/end_draw().

### CODE REVIEW

File: libnodegl/rendertarget_gl.c [hunk 1]
```diff
@@ -81,9 +79,10 @@ static void resolve_draw_buffers(struct rendertarget *s)
 #if defined(TARGET_DARWIN)
         ngli_glDrawBuffer(gl, GL_COLOR_ATTACHMENT0 + i);
 #else
+        GLenum draw_buffers[NGLI_MAX_COLOR_ATTACHMENTS] = {0};
+        draw_buffers[i] = GL_COLOR_ATTACHMENT0 + i;
         ngli_glDrawBuffers(gl, i + 1, draw_buffers);
 #endif
```
Comment (cboesch-gpsw) [inline libnodegl/rendertarget_gl.c:85]: do we still need to call `glDrawBuffers()` just for one buffer?

File: libnodegl/rendertarget_gl.c [hunk 2]
```diff
@@ -397,3 +397,55 @@ void ngli_rendertarget_gl_freep(struct rendertarget **sp)
 
     ngli_freep(sp);
 }
+
+int ngli_default_rendertarget_gl_init(struct rendertarget *s, const struct rendertarget_params *params)
```
Comment (cboesch-gpsw) [inline libnodegl/rendertarget_gl.c:401]: The function name in the commit message doesn't match exactly

File: libnodegl/gctx_gl.h [hunk 3]
```diff
@@ -55,7 +55,7 @@ struct gctx_gl {
     struct texture *color;
     struct texture *ms_color;
     struct texture *depth;
-    /* Capture offscreen render target */
+    /* Capture callback and resources */
```
Comment (cboesch-gpsw) [inline libnodegl/gctx_gl.h:58]: Do we support non-offscreen capture? If not, I'd keep that qualifier (as first word probably: "Offscreen capture callback and resources"). Or do you mean the fields below are used for "[offscreen capture] + [off/on screen resources for related usages]"?

File: libnodegl/gctx_gl.c [hunk 4]
```diff
@@ -305,6 +336,15 @@ static int gl_resize(struct gctx *s, int width, int height, const int *viewport)
     if (ret < 0)
         return ret;
 
+    s_priv->rt->width = gl->width;
+    s_priv->rt->height = gl->height;
+
+    /* The default framebuffer id can change after a resize operation on EAGL,
```
Comment (cboesch-gpsw) [inline libnodegl/gctx_gl.c:342]: Style nit: one extra line break for consistency with the trail:
```c
/*
 * bla bla
 * bla bla
 */
```

File: libnodegl/node_rtt.c [hunk 5]
```diff
@@ -427,7 +429,7 @@ static void rtt_draw(struct ngl_node *node)
     }
     ngli_gctx_end_render_pass(gctx);
 
-    ctx->current_rendertarget = prev_rendertargets[1];
+    ctx->current_rendertarget = prev_rendertargets[current_rendertarget_index];
```
Comment (cboesch-gpsw) [inline libnodegl/node_rtt.c:432]: Comment on the message of this commit:

- In "the RTT 1 node make the rendertarget with LOAD_OP=CLEAR current", "make ... current" → "make*s* current"
- In "the RTT 2 node ends the previous render pass if needed"; what does "if needed" means in this context? does it, or does it not in the scenario you describe?

File: tests/rtt.py [hunk 6]
```diff
@@ -200,3 +200,41 @@ def rtt_load_attachment(cfg):
...
+def rtt_clear_attachment_with_timeranges(cfg):
+    cfg.aspect_ratio = (1, 1)
+
+    quad = ngl.Quad((-1, -1, 0), (2, 0, 0), (0, 2, 0))
+    program = ngl.Program(vertex=cfg.get_vert('color'), fragment=cfg.get_frag('color'))
+    program.update_vert_out_vars(var_tex0_coord=ngl.IOVec2(), var_uvcoord=ngl.IOVec2())
+    render = ngl.Render(quad, program)
+    render.update_frag_resources(color=ngl.UniformVec4(value=COLORS['white']))
+    time_range_filter = ngl.TimeRangeFilter(render)
+    time_range_filter.add_ranges(ngl.TimeRangeModeNoop(0))
```
Comment (cboesch-gpsw) [inline tests/rtt.py:216]: Please add a label and/or a comment above this block: `# Time-disabled full screen white rectangle` or something like that.

The quad and program can be re-used by the way.

File: tests/rtt.py [hunk 7]
```diff
@@ -200,3 +200,41 @@ def rtt_load_attachment(cfg):
...
+    texture = ngl.Texture2D(width=32, height=32)
+    rtt_noop = ngl.RenderToTexture(ngl.Identity(), [texture])
+
+    quad = ngl.Quad((-0.5, -0.5, 0), (1, 0, 0), (0, 1, 0))
+    program = ngl.Program(vertex=cfg.get_vert('color'), fragment=cfg.get_frag('color'))
+    program.update_vert_out_vars(var_tex0_coord=ngl.IOVec2(), var_uvcoord=ngl.IOVec2())
+    render = ngl.Render(quad, program)
+    render.update_frag_resources(color=ngl.UniformVec4(value=COLORS['orange']))
+    animkf = [ngl.AnimKeyFrameFloat(0, 0), ngl.AnimKeyFrameFloat(cfg.duration, -360)]
+    render = ngl.Rotate(render, anim=ngl.AnimatedFloat(animkf))
```
Comment (cboesch-gpsw) [inline tests/rtt.py:227]: `# Centered rotating quad`

File: tests/rtt.py [hunk 8]
```diff
@@ -200,3 +200,41 @@ def rtt_load_attachment(cfg):
...
+    quad = ngl.Quad((-1, -1, 0), (2, 0, 0), (0, 2, 0))
+    program = ngl.Program(vertex=cfg.get_vert('color'), fragment=cfg.get_frag('color'))
+    program.update_vert_out_vars(var_tex0_coord=ngl.IOVec2(), var_uvcoord=ngl.IOVec2())
+    render = ngl.Render(quad, program)
+    render.update_frag_resources(color=ngl.UniformVec4(value=COLORS['white']))
+    time_range_filter = ngl.TimeRangeFilter(render)
+    time_range_filter.add_ranges(ngl.TimeRangeModeNoop(0))
+
+    texture = ngl.Texture2D(width=32, height=32)
+    rtt_noop = ngl.RenderToTexture(ngl.Identity(), [texture])
```
Comment (cboesch-gpsw) [inline tests/rtt.py:219]: `# Intermediate no-op RTT to force a new render pass internally`

File: tests/rtt.py [hunk 9]
```diff
@@ -200,3 +200,41 @@ def rtt_load_attachment(cfg):
...
+    program.update_vert_out_vars(var_tex0_coord=ngl.IOVec2(), var_uvcoord=ngl.IOVec2())
+    render = ngl.Render(quad, program)
+    render.update_frag_resources(color=ngl.UniformVec4(value=COLORS['orange']))
+    animkf = [ngl.AnimKeyFrameFloat(0, 0), ngl.AnimKeyFrameFloat(cfg.duration, -360)]
+    render = ngl.Rotate(render, anim=ngl.AnimatedFloat(animkf))
+
+    group = ngl.Group(children=(time_range_filter, rtt_noop, render))
+
+    texture = ngl.Texture2D(width=512, height=512)
+    rtt = ngl.RenderToTexture(group, [texture])
```
Comment (cboesch-gpsw) [inline tests/rtt.py:232]: `# Root RTT`

File: tests/rtt.py [hunk 10]
```diff
@@ -200,3 +200,41 @@ def rtt_load_attachment(cfg):
...
+
+    quad = ngl.Quad((-1, -1, 0), (2, 0, 0), (0, 2, 0))
+    program = ngl.Program(vertex=cfg.get_vert('texture'), fragment=cfg.get_frag('texture'))
+    program.update_vert_out_vars(var_tex0_coord=ngl.IOVec2(), var_uvcoord=ngl.IOVec2())
+    render = ngl.Render(quad, program)
+    render.update_frag_resources(tex0=texture)
```
Comment (cboesch-gpsw) [inline tests/rtt.py:238]: `# Full screen rendering of the root RTT result`

File: libnodegl/gctx_gl.c [hunk 11]
```diff
@@ -196,7 +196,40 @@ static int offscreen_rendertarget_init(struct gctx *s)
     return 0;
 }
 
-static void offscreen_rendertarget_reset(struct gctx *s)
+static int onscreen_rendertarget_init(struct gctx *s)
+{
+    struct gctx_gl *s_priv = (struct gctx_gl *)s;
+    struct ngl_config *config = &s->config;
```
Comment (mbouron-gpsw) [inline libnodegl/gctx_gl.c:202]: const.

File: libnodegl/rendertarget_gl.c (hunk 1 above)
Comment (mbouron-gpsw) [reply libnodegl/rendertarget_gl.c:85]: If there is only one color attachment, this implementation is not called (resolve_no_draw_buffers() is called instead).

File: libnodegl/rendertarget_gl.c (hunk 2 above)
Comment (mbouron-gpsw) [reply libnodegl/rendertarget_gl.c:401]: Fixed.

File: libnodegl/gctx_gl.h (hunk 3 above)
Comment (mbouron-gpsw) [reply libnodegl/gctx_gl.h:58]: Reworded to Offscreen capture callback and resources.

File: libnodegl/gctx_gl.c (hunk 4 above)
Comment (mbouron-gpsw) [reply libnodegl/gctx_gl.c:342]: Fixed.

File: libnodegl/rendertarget_gl.c (hunk 1 above)
Comment (mbouron-gpsw) [reply libnodegl/rendertarget_gl.c:85]: And yes we need to call glDrawBuffers(), it's the only way to select a specific color attachment on OpenGLES. Moreover, the index of the draw buffers must match the index of the color attachment, ie: i == GL_COLOR_ATTACHMENT0 + i.

File: libnodegl/node_rtt.c (hunk 5 above)
Comment (mbouron-gpsw) [reply libnodegl/node_rtt.c:432]: - Fixed
- Removed this line, and explained in the previous step that the render pass associated with the current rendertarget is not started

File: tests/rtt.py (hunk 10 above)
Comment (mbouron-gpsw) [reply tests/rtt.py:238]: Done.

File: tests/rtt.py (hunk 9 above)
Comment (mbouron-gpsw) [reply tests/rtt.py:232]: Done.

File: tests/rtt.py (hunk 8 above)
Comment (mbouron-gpsw) [reply tests/rtt.py:219]: Done.

File: tests/rtt.py (hunk 7 above)
Comment (mbouron-gpsw) [reply tests/rtt.py:227]: Done.

File: tests/rtt.py (hunk 6 above)
Comment (mbouron-gpsw) [reply tests/rtt.py:216]: Done, I didn't reuse the quad though.

**Review by cboesch-gpsw — APPROVED** [review]

**Pull request merged**